File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/EndOfLineFilter.java
package com.codepoetics.raffia.streaming;

import com.codepoetics.raffia.writers.BasketWriter;

import java.math.BigDecimal;

public final class EndOfLineFilter<T extends BasketWriter<T>> implements FilteringWriter<T> {

  private final T target;

  public EndOfLineFilter(T target) {
    this.target = target;
  }

  @Override
  public FilteringWriter<T> beginObject() {
    throw new IllegalStateException("beginObject() called after basket complete");
  }

  @Override
  public FilteringWriter<T> beginArray() {
    throw new IllegalStateException("beginArray() called after basket complete");
  }

  @Override
  public FilteringWriter<T> end() {
    throw new IllegalStateException("end() called after basket complete");
  }

  @Override
  public FilteringWriter<T> key(String key) {
    throw new IllegalStateException("key() called after basket complete");
  }

  @Override
  public FilteringWriter<T> add(String value) {
    throw new IllegalStateException("add() called after basket complete");
  }

  @Override
  public FilteringWriter<T> add(BigDecimal value) {
    throw new IllegalStateException("add() called after basket complete");
  }

  @Override
  public FilteringWriter<T> add(boolean value) {
    throw new IllegalStateException("add() called after basket complete");
  }

  @Override
  public FilteringWriter<T> addNull() {
    throw new IllegalStateException("addNull() called after basket complete");
  }

  @Override
  public FilteringWriter<T> advance(T newTarget) {
    throw new IllegalStateException("advance() called after basket complete");
  }

  @Override
  public T complete() {
    return target;
  }
}

File: raffia-core/src/main/java/com/codepoetics/raffia/streaming/EndOfLineFilter.kt
package com.codepoetics.raffia.streaming

import com.codepoetics.raffia.writers.BasketWriter

import java.math.BigDecimal

class EndOfLineFilter<T : BasketWriter<T>>(private val target: T) : FilteringWriter<T> {

    override fun beginObject(): FilteringWriter<T> {
        throw IllegalStateException("beginObject() called after basket complete")
    }

    override fun beginArray(): FilteringWriter<T> {
        throw IllegalStateException("beginArray() called after basket complete")
    }

    override fun end(): FilteringWriter<T> {
        throw IllegalStateException("end() called after basket complete")
    }

    override fun key(key: String): FilteringWriter<T> {
        throw IllegalStateException("key() called after basket complete")
    }

    override fun add(value: String): FilteringWriter<T> {
        throw IllegalStateException("add() called after basket complete")
    }

    override fun add(value: BigDecimal): FilteringWriter<T> {
        throw IllegalStateException("add() called after basket complete")
    }

    override fun add(value: Boolean): FilteringWriter<T> {
        throw IllegalStateException("add() called after basket complete")
    }

    override fun addNull(): FilteringWriter<T> {
        throw IllegalStateException("addNull() called after basket complete")
    }

    override fun advance(newTarget: T): FilteringWriter<T> {
        throw IllegalStateException("advance() called after basket complete")
    }

    override fun complete(): T {
        return target
    }
}
